eth2/proto_array_fork_choice/src/proto_array.rs

```diff
@@ -0,0 +1,417 @@
+    ///
+    /// ## Notes
+    ///
+    /// The result of this function is not guaranteed to be accurate if `Self::on_new_block` has
+    /// been called without a subsequent `Self::apply_score_changes` call. This is because
+    /// `on_new_block` does not attempt to walk backwards through the tree and update the
+    /// best-child/best-descendant links.
+    pub fn find_head(&self, justified_root: &Hash256) -> Result<Hash256, Error> {
+        let justified_index = self
+            .indices
+            .get(justified_root)
+            .copied()
+            .ok_or_else(|| Error::JustifiedNodeUnknown(*justified_root))?;
+
+        let justified_node = self
+            .nodes
+            .get(justified_index)
+            .ok_or_else(|| Error::InvalidJustifiedIndex(justified_index))?;
+
+        let best_descendant_index = justified_node
+            .best_descendant
+            .unwrap_or_else(|| justified_index);
+
+        let best_node = self
+            .nodes
+            .get(best_descendant_index)
+            .ok_or_else(|| Error::InvalidBestDescendant(best_descendant_index))?;
+
+        // It is a logic error to try and find the head starting from a block that does not match
+        // the filter.
+        if !self.node_is_viable_for_head(&best_node) {
+            return Err(Error::InvalidBestNode {
+                justified_epoch: self.justified_epoch,
+                finalized_epoch: self.finalized_epoch,
+                node_justified_epoch: justified_node.justified_epoch,
+                node_finalized_epoch: justified_node.finalized_epoch,
+            });
+        }
+
+        // dbg!(&self.nodes);
+
+        Ok(best_node.root)
+    }
+
+    /// Update the tree with new finalization information. The tree is only actually pruned if both
+    /// of the two following criteria are met:
+    ///
+    /// - The supplied finalized epoch and root are different to the current values.
+    /// - The number of nodes in `self` is at least `self.prune_threshold`.
+    ///
+    /// # Errors
+    ///
+    /// Returns errors if:
+    ///
+    /// - The finalized epoch is less than the current one.
+    /// - The finalized epoch is equal to the current one, but the finalized root is different.
+    /// - There is some internal error relating to invalid indices inside `self`.
+    pub fn maybe_prune(
+        &mut self,
+        finalized_epoch: Epoch,
+        finalized_root: Hash256,
+    ) -> Result<(), Error> {
+        if finalized_epoch < self.finalized_epoch {
+            // It's illegal to swap to an earlier finalized root (this is assumed to be reverting a
+            // finalized block).
+            return Err(Error::RevertedFinalizedEpoch {
+                current_finalized_epoch: self.finalized_epoch,
+                new_finalized_epoch: finalized_epoch,
+            });
+        } else if finalized_epoch != self.finalized_epoch {
+            self.finalized_epoch = finalized_epoch;
+        }
+
```

Comment:
Do we need to check the case where finalized epoch is equal to current one but finalized root is different? I don't see the check here

Reply:
Yeah, that would be a good idea. I'm still debugging some of this finality stuff at the moment.

You might find this interesting: https://github.com/ethereum/eth2.0-specs/issues/1574